### PR TITLE
Correctly change the working directory in python

### PR DESCRIPTION
**Pull request by Leandros**
This is implementing #295 and https://github.com/neovim/neovim/issues/3636.

### CODE REVIEW

**Comment by Leandros** [comment]
@justinmk Sure, I've had a test written anyway.

File: neovim/plugin/script_host.py [hunk 1]
```diff
@@ -65,7 +65,9 @@ def setup(self, nvim):
         sys.stderr = RedirectStream(lambda data: nvim.err_write(data))
         # Setup the DirChanged listener
         cid = nvim.channel_id
-        nvim.command('au DirChanged * call rpcrequest(%d, "python_chdir", v:event)' % cid)
+        cmd = 'au DirChanged * call rpcrequest(%d, "python_chdir", v:event)' \
```
Comment (mhinz) [inline neovim/plugin/script_host.py]: Please use `autocmd` instead of `au`, which makes it easier go grep for when searching for autocmd-related things.

Comment (Leandros) [reply neovim/plugin/script_host.py]: Sure.

**Comment by Leandros** [comment]
Alright, all tests ran through. Fixed the last linter warning.

File: neovim/plugin/script_host.py [hunk 2]
```diff
@@ -63,6 +63,12 @@ def setup(self, nvim):
         self.saved_stderr = sys.stderr
         sys.stdout = RedirectStream(lambda data: nvim.out_write(data))
         sys.stderr = RedirectStream(lambda data: nvim.err_write(data))
+        # Setup the DirChanged listener
```
Comment (bfredl) [inline neovim/plugin/script_host.py]: Hmm maybe we can use the host autocmd infrastructure here. (We need to add a flag to not start the host for the autocmd, but that would be useable in general)

Comment (bfredl) [reply neovim/plugin/script_host.py]: Actually that would require a bit bigger refactor than I thought, so let's keep this as is for now (and maybe revisit at `UpdateRemotePlugins` refactor)

**Comment by mhinz** [comment]
@bfredl 

What is happening here? https://travis-ci.org/neovim/python-client/jobs/330770820#L512

I get this locally as well. Sometimes.

```
Error detected while processing function provider#python3#Call:
line   18:
no request handler registered for "python_execute"
```

It starts happening at a8a98b8.

**Comment by bfredl** [comment]
you mean you don't see it on master?

File: neovim/plugin/script_host.py [hunk 3]
```diff
@@ -63,6 +63,12 @@ def setup(self, nvim):
...
+        cid = nvim.channel_id
```
Comment (bfredl) [inline neovim/plugin/script_host.py]: unnecessary temporary

File: neovim/plugin/script_host.py [hunk 4]
```diff
@@ -63,6 +63,12 @@ def setup(self, nvim):
...
+        cmd = 'autocmd DirChanged *' \
+            + ' call rpcrequest(%d, "python_chdir", v:event)' \
```
Comment (bfredl) [inline neovim/plugin/script_host.py]: also we prefer to use new-style `.format()`

**Comment by mhinz** [comment]
Yup. I just do:

```
$ git checkout a8a98b8
$ nvim
:python3 import os
```

Commenting out the following line makes it work again:

```
nvim.command('au DirChanged * call rpcrequest(%d, "python_chdir", v:event)' % cid)
```

**Comment by mhinz** [comment]
Any `nvim.command()` in `setup()` will lead to that error.

**Comment by bfredl** [comment]
Hmm, maybe it shouldn't be done in  `__init__`. I generally recommend against doing any RPC there, but I don't  see the specific problem right here. Maybe the issue is that it is un-synchorized in general, using `nvim.command(...,async=True)` might fix it even if its not pretty either.

**Comment by mhinz** [comment]
Yup,`nvim.command(..., async=True)` fixes it for me.

File: neovim/plugin/script_host.py (hunk 4 above)
Comment (Leandros) [reply neovim/plugin/script_host.py]: I, too, prefer `.format()`. I simply followed what I've seen being used before. Will change it.

**Comment by bfredl** [comment]
It is a bit racy, if a vim script does
```
py3 import blah # first call, so starts the host
cd mydir
```
the `cd` might come before the async registration. One can of course also invoke it in the async command to be sure in this case, but generally we want some mechanism for "run this vimL code right when the host starts".

**Comment by Leandros** [comment]
@mhinz Hmm. That is interesting. I'm using it already, and couldn't see any errors. Which nvim version are you using?

**Comment by mhinz** [comment]
@Leandros 

Hm, you're right. I can only reproduce it with master and this repo is using nightlies during CI tests.

I can't reproduce it with the 0.2.2 release.

I'll bisect nvim later.

**Comment by mhinz** [comment]
@bfredl 

This issue starts happening after https://github.com/neovim/neovim/commit/1ebc96fe1.

**Comment by bfredl** [comment]
Yes, it changes processing of RPC job messages,  but not more than to how they were before RPC jobs gained stderr support neovim/neovim#4723 (and to how RPC sockets always have worked), so they are less "forgiving" to potentially racy patterns.

Synchronous initialization should be initiated by one party only. In the case of plugin hosts, this is already nvim, by using one of `poll`, `specs` or in the case of `script_host` by calling one of the script methods directly.

File: neovim/plugin/script_host.py [hunk 5]
```diff
@@ -44,6 +44,14 @@ def __init__(self, nvim):
         exec('import sys', self.module.__dict__)
         self.legacy_vim = LegacyVim.from_nvim(nvim)
         sys.modules['vim'] = self.legacy_vim
+        # Since the 'DirChanged' autocmd is established asynchronously, import
+        # os manually already, to prevent a potential race condition
+        exec('import os', self.module.__dict__)
```
Comment (bfredl) [inline neovim/plugin/script_host.py]: Why is this needed? The autocmd handler method is executed in the context of `script_host.py` which already has the import.

Comment (Leandros) [reply neovim/plugin/script_host.py]: I tried to work-around the race you described, by importing os before, so that Vim can see the current working directory without creating a race.

**Comment by Leandros** [comment]
@bfredl What's your suggestion of implementing this? It's definitely required, we need to watch for the DirChange somehow. I've just looked at the ruby plugin, and found they're doing it quite similar.

File: neovim/plugin/script_host.py (hunk 5 above)
Comment (bfredl) [reply neovim/plugin/script_host.py]: But the race has nothing to do with `import os`. Quick fix would rather be to do 
`nvim.command('call rpcrequest({}, "python_chdir", {"cwd": getcwd()}', async=True)` after registering the autocmd

**Comment by bfredl** [comment]
Lets use the current solution for now (with the fix I just posted), until we improve host autocmd registration.

File: neovim/plugin/script_host.py (hunk 5 above)
Comment (Leandros) [reply neovim/plugin/script_host.py]: The quickfix doesn't seem to work, at least for me, it results in a python error from my plugins.
screen shot 2018-01-24 at 10 11 03

**Comment by justinmk** [comment]
@Leandros can you include the fix @bfredl mentioned? Would like to move this forward

**Comment by justinmk** [comment]
> `nvim.command('call rpcrequest({}, "python_chdir", {"cwd": getcwd()}', async=True)`

seems to be a chicken-egg problem, instead I went with this:

    os.chdir(nvim.eval('getcwd()', async=False))

Pushed an update to @Leandros branch, will merge after CI.

**Comment by blueyed** [comment]
This appears to cause problems with initial use of `:py3` for me: https://github.com/neovim/python-client/issues/337.
